### PR TITLE
Add DEPLOYER_IMAGE x-google-marketplace property.

### PR DESCRIPTION
**Pull request by trironkk**
Paired marketplace-k8s-app-example pull request (https://github.com/GoogleCloudPlatform/marketplace-k8s-app-example/pull/110).

This is a follow up task in https://github.com/GoogleCloudPlatform/marketplace-k8s-app-tools/issues/242.

### CODE REVIEW

**Comment by trironkk** [comment]
Will also update our backend to be consistent with `provision.py`.

File: marketplace/deployer_util/config_helper.py [hunk 1]
```diff
@@ -201,6 +202,9 @@ def __init__(self, name, dictionary, required):
       elif xt == XTYPE_IMAGE:
         d = self._x.get('image', {})
         self._image = SchemaXImage(d)
+      elif xt == XTYPE_DEPLOYER_IMAGE:
+        d = self._x.get('image', {})
```
Comment (huyhg) [inline marketplace/deployer_util/config_helper.py]: s/image/deployerImage/

File: marketplace/deployer_util/provision.py [hunk 2]
```diff
@@ -53,6 +53,16 @@ def process(schema, values, deployer_image, deployer_entrypoint):
   app_name = get_name(schema, values)
   namespace = get_namespace(schema, values)
 
+  # Inject DEPLOYER_IMAGE property values if not already present.
+  for key, value in values.items():
```
Comment (huyhg) [inline marketplace/deployer_util/provision.py]: I believe this has a bug.

We need to iterate over `schema.properties` and fine the `DEPLOYER_IMAGE` propertis. `values` wouldn't have that key. Then:

```
if key not in values:
  values[key] = deployer_image
```

File: marketplace/deployer_util/provision.py [hunk 3]
```diff
@@ -53,6 +53,16 @@ def process(schema, values, deployer_image, deployer_entrypoint):
...
+    if value:
+      continue
+    if key not in schema.properties:
+      continue
+    if not schema.properties[key].xtype == 'DEPLOYER_IMAGE':
+      continue
+    values[key] = deployer_image
```
Comment (huyhg) [inline marketplace/deployer_util/provision.py]: can you extract this into a function and add a test case? This file is missing a lot of tests, but I think we can test this particular thing.

Comment (trironkk) [reply marketplace/deployer_util/provision.py]: Added a test, though I'm unsatisfied with it.

I'd be happy with a test that validated that `schema.yaml` + `values.yaml` -> `manifest.yaml`, but there's other higher-impact broader-scope tests that would add more value (e.g. sample deployers for each of the base deployers).

File: marketplace/deployer_util/config_helper_test.py [hunk 4]
```diff
@@ -49,6 +49,11 @@
     default: gcr.io/google/busybox
     x-google-marketplace:
       type: IMAGE
+  propertyDeployerImage:
+    type: string
+    default: gcr.io/google/deployer
```
Comment (huyhg) [inline marketplace/deployer_util/config_helper_test.py]: rm default value

File: marketplace/deployer_util/provision.py [hunk 5]
```diff
@@ -53,6 +53,9 @@ def process(schema, values, deployer_image, deployer_entrypoint):
   app_name = get_name(schema, values)
   namespace = get_namespace(schema, values)
 
+  # Inject DEPLOYER_IMAGE property values if not already present.
+  values = inject_deployer_image_properties(schema, deployer_image, values)
```
Comment (huyhg) [inline marketplace/deployer_util/provision.py]: optional:
- remove the return value to make the API cleaner: it's modifying the passed in values directly
- Make `values` the first argument

File: marketplace/deployer_util/provision_test.py [hunk 6]
```diff
@@ -43,3 +46,17 @@ def test_limit_name(self):
...
+    properties:
+      deployer_image:
+        type: string
+        x-google-marketplace:
+          type: DEPLOYER_IMAGE
+    ''')
+    values = {}
+    deployer_image = 'gcr.io/cloud-marketplace/partner/solution/deployer:latest'
+    self.assertEquals(
+        provision.inject_deployer_image_properties(
```
Comment (huyhg) [inline marketplace/deployer_util/provision_test.py]: removing the return value will inconvenience the test--well I'm not opposed to having the method returning the input `values` for this purpose

**Pull request merged**